homeassistant/components/tplink/light.py

```diff
@@ -73,7 +73,15 @@ async def async_setup_entry(hass: HomeAssistantType, config_entry, async_add_ent
 def get_devices_sysinfo(devices):
     """Get sysinfo for all devices."""
     for device in devices:
-        device.get_sysinfo()
+        try:
+            device.get_sysinfo()
+        except SmartDeviceException as ex:
+            _LOGGER.warning(
+                "Unable to communicate with device %s due to: %s",
+                device.host,
+                ex,
+            )
+            devices.remove(device)
```

Comment:
You can't remove an item from a list while iterating over it. 

It also seems weird that we would completely remove a device if we can't reach it at startup.

Instead, this check needs to be moved into `async_update` of the entity and entity should be marked as unavailable if this info couldn't be detected.

Reply:
Just want to be clear, is there a way of adding the entity if I don't yet have the ID? Obtaining the ID is currently dependent on running that step. Otherwise we only have a static IP which in some cases could no longer be valid (if it was DHCP for example). I also thought of moving that to the ```async_update``` but I am not clear how to handle the ID otherwise. If I do change the logic on the ID generation I assume that would also become a breaking change.

Reply:
You can decide to add the entity later when it becomes available by keeping `async_add_entities` around and retrying for the devices that failed.

Reply:
Oh I see that's what you did